ema_workbench/em_framework/optimization.py

```diff
@@ -572,25 +808,44 @@ def __init__(self, metrics, max_nfe, convergence_freq=1000, logging_freq=5, log_
         self.convergence_freq = convergence_freq
         self.logging_freq = logging_freq
 
+        # TODO what is the point of this code?
         for metric in metrics:
             assert isinstance(metric, AbstractConvergenceMetric)
             metric.reset()
 
-    def __call__(self, optimizer):
+    def __call__(self, optimizer, force=False):
+        """Stores convergences information given specified convergence
+        frequency.
+
+        Parameters
+        ----------
+        optimizer : platypus optimizer instance
+        force : boolean, optional
+                if True, convergence information will always be stored
+                if False, converge information will be stored if the
+                the number of nfe since the last time of storing is equal to
+                or higher then convergence_freq
+
+
+        the primary use case for force is to force convergence frequency information
+        to be stored once the stopping condition of the optimizer has been reached
+        so that the final convergence information is kept.
+
+        """
         nfe = optimizer.nfe
         super().__call__(nfe - self.i)
 
         self.generation += 1
 
-        if (nfe >= self.last_check + self.convergence_freq) or self.last_check == 0:
+        if (nfe >= self.last_check + self.convergence_freq) or (self.last_check == 0) or force:
             self.index.append(nfe)
             self.last_check = nfe
 
             for metric in self.metrics:
                 metric(optimizer)
 
     def to_dataframe(self):
-        progress = {metric.name: metric.results for metric in self.metrics if metric.results}
+        progress = {metric.name: result for metric in self.metrics if result := metric.get_results()}
```

Comment:
I forgot what a bitch assignment expressions was [about parentheses](https://peps.python.org/pep-0572/#exceptional-cases). This should fix the syntax error and CI issues.

```suggestion
        progress = {metric.name: result for metric in self.metrics if (result := metric.get_results())}
```

Reply:
@quaquel I will just commit this so we can see if the CI goes green again!